Fix shader baker metal crash

TCROC opened the pull request:
Fixes https://github.com/godotengine/godot/issues/109846

This passes the ``ExportPreset`` to the shader baker so that the it can take into account the export settings when compiling the shaders.  For this particular PR, it passes ``-mios-version-min=`` and ``-mmacosx-version-min=`` to ``xcrun``.  ``xcrun`` is fortunately smart enough to know which version of metal to compile the shaders against when we pass it ``application/min_macos_version_arm64`` and ``application/min_ios_version`` from the export preset.

TCROC commented:
I'm still running some tests to make sure this covers all cases.  As of the moment, this seems to fix things!  I'd love to hear your thoughts on this @stuartcarnie! :)

Also shoutout to @stuartcarnie for all the help on narrowing down the issue!  You were a wizard in making sense of my logs!

TCROC commented:
Looks like some files were moved around. I'll update the headers tomorrow. Shouldn't effect any of the logic tho so still good for feedback

DarioSamo commented:
- Agree with bruvzg's points.
- You'll need to find a way to affect the customization hash as well. This will probably require extending the customization hash function to be affected by the current platform implementation. Otherwise, your shaders won't be recreated if you just change the options like the minimum OS version and export again, leading to a frustrating issue to debug when changing the export option doesn't fix it.

TCROC commented:
Ah that makes sense. I'll go take a look at both of those suggestions

TCROC committed:
Fix shader baker targetting higher than min supported os in export_presets.cfg

TCROC commented on drivers/metal/rendering_shader_container_metal.h:
```diff
@@ -222,7 +223,7 @@ class RenderingShaderContainerMetal : public RenderingShaderContainer {
 	static constexpr uint32_t FORMAT_VERSION = 1;
 
-	void set_export_mode(bool p_export_mode) { export_mode = p_export_mode; }
+	void set_export_preset(const Ref<EditorExportPreset> &p_preset) { preset = p_preset; }
```
Done :)

TCROC commented:
@DarioSamo Let me know what you think of the implementation! :). I added a ``get_customization_configuration_info`` method to the ``rendering_shader_container`` so that custom shader containers can override it and provide the baker with information for generating the shader hashes.

stuartcarnie commented:
Hey @TCROC, great start on this PR, however, it still suffers from the same problem I previously introduced where the compiled .metallib version would mismatch the serialised version in the baked shader header. I've opened #110264 to use the `-std=metalX.X` argument to guarantee the compiled version of the `.metallib` matches what we serialise, so that when we perform the version check at runtime, we can return an error and the Godot renderer will recompile the shader.

I think the direction you've headed is good, as user's should only have to think about their minimum target OS. However, we'll need to create a table that maps OS version to Metal version, so we can still ensure the compiled `.metallib` language version is known to the serializer / baker to be able to perform the runtime checks and return an error if the target OS has a lower Metal language requirement.

stuartcarnie commented:
> but then it threw an error that it needed a 2.x version.

Is the device different from the logs you shared in [this comment](https://github.com/godotengine/godot/issues/109846#issuecomment-3249672929), as that device is an A15 (Apple8 GPU) which is definitely a Metal 3 (and 4) device according to Apple's feature set tables:

<img width="40%" alt="CleanShot 2025-09-05 at 15 03 01@2x" src="https://github.com/user-attachments/assets/f6d4cc26-7a74-4f75-9a83-0d1096bb4bbc" />

Are you also testing on another device that has lower hardware specs?

It may be that the issue is determining the device's maximum Metal version. We use `MTLCompileOptions.languageVersion` to determine the maximum language version of the device. It is possible that this only reports the maximum of the OS rather than the hardware. 

> Is there a way to detect the error when "main0" is called and then recompile from there? This would be the most reliable solution as Apple's internal logic will tell us if it is correct or not.

We load the library in the function that parses the shader container, but that clearly does not generate an error. It only fails when it attempts to compile the pipeline, which is currently too late in the Godot render path. 

Could you try specifying `-std=metal3.1` _and_ the `--ios-version-min` and see if that works? I ask as this log you shared suggests it might still be using `metal3.1` along with the default OS target of your compiler, which is a newer iOS:

```
Function main0 has a deployment target (0x00020007) which is incompatible with this OS (0x00020006).
```

This would ensure we know the language version to be `3.1` and also correctly targeting an earlier OS version.

TCROC commented:
> Is the device different from the logs you shared in https://github.com/godotengine/godot/issues/109846#issuecomment-3249672929, as that device is an A15 (Apple8 GPU) which is definitely a Metal 3 (and 4) device according to Apple's feature set tables:

So this is where it gets confusing.  The device is the same.  It **IS** a Metal 3 / 4 device but **IF AND ONLY IF** it is running iOS 18.  On iOS 17 it is metal 2.  At least that is what my tests are showing on this device.  (Also, not yelling.  Just trying to emphasize the conditions that result in certain metal versions being required as it is very confusing and unclear imo).

> Are you also testing on another device that has lower hardware specs?

Nope.  The same device.

> It may be that the issue is determining the device's maximum Metal version. We use MTLCompileOptions.languageVersion to determine the maximum language version of the device. It is possible that this only reports the maximum of the OS rather than the hardware.

This definitely could be the case.  Especially since the issue appears to be a hardware + software combination that determines the actual max metal version.

> We load the library in the function that parses the shader container, but that clearly does not generate an error. It only fails when it attempts to compile the pipeline, which is currently too late in the Godot render path.

Ah I see.  I'm personally much more familiar with writing in Rust.  In Rust there is a "sender / receiver" pattern built in where I want send messages across threads.  But its also useful for just sending information across different stages of a program if a back track is needed in order to handle an error.  How challenging would it be to communicate something back up the render path that "some really bad stuff is going on. lets try this again without the shader baker."?

> Could you try specifying -std=metal3.1 and the --ios-version-min and see if that works? I ask as this log you shared suggests it might still be using metal3.1 along with the default OS target of your compiler, which is a newer iOS. This would ensure we know the language version to be 3.1 and also correctly targeting an earlier OS version.

Yep!  I'll have access to the device today and can run these tests!

Thanks for the feedback!  I look forward to taking a look at these!  I'll report back with what I find :)

stuartcarnie commented:
> Ah I see. I'm personally much more familiar with writing in Rust. In Rust there is a "sender / receiver" pattern built in where I want send messages across threads. But its also useful for just sending information across different stages of a program if a back track is needed in order to handle an error. How challenging would it be to communicate something back up the render path that "some really bad stuff is going on. lets try this again without the shader baker."?

I write Rust, and have written a lot of Go, and both have used channels extensively for communicating sequential processes. C++ can do that too, but it's only relevant if the software architecture is orchestrated in such a way to leverage these primitives. The renderer has several phases that take place:

1. Perform background compilation of all the shader variants, which includes pulling them from the baked cache, runtime cache and / or recompiling from original GLSL source if there is an issue or missing.
2. Compile specific pipelines when needed to render a given scene

The issue is that phase 1 happens and Godot may never use the shaders in a compiled pipeline, so nothing more happens with them. Channels wouldn't help here, because it wasn't the renderer asking for a shader to begin with. 

The issue is that for Metal, we _load_ the `MTLLibrary` binary object, which appears to succeed regardless of whether the shader can be used by a Metal pipeline for the current OS. Only when the pipeline is compiled, when the shader is about to be used, does it fail. This happens well after initialisation of Godot, and so there is no back channel, etc. The code requesting it compile the shader may not even have the original source it requested, as it has a reference to a shader object, like [`RDShaderSPIRV`](https://docs.godotengine.org/en/stable/classes/class_rdshaderspirv.html)[^1]

> So this is where it gets confusing. The device is the same. It IS a Metal 3 / 4 device but IF AND ONLY IF it is running iOS 18. On iOS 17 it is metal 2. At least that is what my tests are showing on this device. (Also, not yelling. Just trying to emphasize the conditions that result in certain metal versions being required as it is very confusing and unclear imo).

Apple8 GPU _should_ be a Metal 3.x capable device on iOS 17 – that is a very new GPU. Metal 3.0 was introduced in iOS 16, so all Metal 3 capable GPUs should be running / supporting Metal 3 on iOS 16+[^2]:

```c++
    MTLLanguageVersion3_0 API_AVAILABLE(macos(13.0), ios(16.0)) = (3 << 16) + 0,
    MTLLanguageVersion3_1 API_AVAILABLE(macos(14.0), ios(17.0)) =  (3 << 16) + 1,
    MTLLanguageVersion3_2 API_AVAILABLE(macos(15.0), ios(18.0)) = (3 << 16) + 2,
```

It's an easy test if you try with a little bit of Swift on the device (and this query was also introduced in iOS 16.0):

```swift
guard let device = MTLCreateSystemDefaultDevice() else {
    println("No Metal available!")
    return
}
if device.supportsFamily(.metal3) {
    println("Metal 3 capable device")
    return
}
println("Metal 2.x capable device")
```

[^1]: Internal to the engine it is not quite this, as it has RIDs, but essentially the same thing.
[^2]: Apple's own feature tables indicate that Apple7+ GPUs are Metal 3 capable, so that means from iOS 16+ they would return `true` for the `supportsFamily(.metal3)` query.

TCROC commented:
Thanks for this insight!  The good news is I'll have access to the device a few hours from now and will be able to run the tests :)

stuartcarnie commented:
Awesome – look forward to hearing what you uncover!

stuartcarnie commented:
@TCROC to clarify, did you specify -std=metal3.1 along with the iOS-min-version=ios17 (or whatever the argument for iOS 17 is)

TCROC commented:
@stuartcarnie That is correct.  I specified them together.  I ran several tests this morning.  The results do not align with Apple's charts.  I'll be getting access to the device again tomorrow to continue building out the chart.  But here it is so far:

### Device specs + app details

```
Incident Identifier: 27568FB0-EA5A-4720-97A9-CD84C03B00F6
Distributor ID:      com.apple.TestFlight
Hardware Model:      iPhone14,7
Process:             blockyballot [1552]
Path:                /private/var/containers/Bundle/Application/20EB61E7-FBBB-4871-9E23-E639F672A1C9/blockyballot.app/blockyballot
Identifier:          com.langestudios.blockyballot
Version:             0.1.43 (1)
AppStoreTools:       16F7
AppVariant:          1:iPhone14,7:17.4
Beta:                YES
Code Type:           ARM-64 (Native)
Role:                Foreground
Parent Process:      launchd [1]
Coalition:           com.langestudios.blockyballot [1125]

Date/Time:           2025-09-02 21:35:42.0867 -0400
Launch Time:         2025-09-02 21:35:39.2242 -0400
OS Version:          iPhone OS 17.6.1 (21G93)
Release Type:        User
Baseband Version:    2.60.02
Report Version:      104
```

### Extra device + godot info in beginning of log

```
blockyballot(4978,0x1f5a1fec0) malloc: Unable to set up reclaim buffer (46) - disabling large cache
Initialize godot-rust (API v4.5.beta.mono.double.custom_build, runtime v4.5.beta.mono.double.custom_build)
Setting up keyboard input view.
Adding observer for keyboard show/hide.
Start animation!
Godot Engine v4.5.beta.mono.double.custom_build.5c05703b6 (2025-09-04 18:02:30 UTC) - https://godotengine.org
TextServer: Added interface "Dummy"
TextServer: Added interface "ICU / HarfBuzz / Graphite (Built-in)"
Devices:
  #0: Apple Apple A15 GPU (Apple8) - Supported, Integrated
- Metal multiview supported:
  max view count: 16
  max instances: 4294967295
Metal 3.1 - Forward Mobile - Using Device #0: Apple - Apple A15 GPU (Apple8)
Using "default" pen tablet driver...
fopen failed for data file: errno = 2 (No such file or directory)
Errors found! Invalidating cache...
Shader 'CanvasSdfShaderRD' (group 0) SHA256: 5195661a4c4664213635bdf5322e7659ec6daf6a99a2b937906435d240996c35
Loading cache for shader CanvasSdfShaderRD, variant 0
Loading cache for shader CanvasSdfShaderRD, variant 1
Loading cache for shader CanvasSdfShaderRD, variant 2
Loading cache for shader CanvasSdfShaderRD, variant 3
Loading cache for shader CanvasSdfShaderRD, variant 4
Loading cache for shader CanvasSdfShaderRD, variant 5
fopen failed for data file: errno = 2 (No such file or directory)
Errors found! Invalidating cache...
Shader 'SkeletonShaderRD' (group 0) SHA256: 0e31747b371ebc780e6f71af9aa0aca6639cf983cf8e223ef73e6499d93ea6fe
Loading cache for shader SkeletonShaderRD, variant 0
Loading cache for shader SkeletonShaderRD, variant 1
Shader 'SortShaderRD' (group 0) SHA256: 1db9a9751100d8052180002c5bc342410e5c1bee3f39104d16eddd55e2510e4a
Loading cache for shader SortShaderRD, variant 0
Loading cache for shader SortShaderRD, variant 1
Loading cache for shader SortShaderRD, variant 2
Warning: Compilation succeeded with: 
```

### Code for compiling xcrun

1. ``std=metalX.Y``

```objc
	// Build the metallib binary.
	{
		List<String> args{ "-sdk", sdk, "metal", "-O3" };

		args.push_back("-std=metal2.6");

		if (p_stage_data.is_position_invariant) {
			args.push_back("-fpreserve-invariance");
		}
		args.push_back("-fmetal-math-mode=fast");
		args.push_back(source_file->get_path_absolute());
		args.push_back("-o");
		args.push_back(result_file->get_path_absolute());
		String r_pipe;
		int exit_code;
		Error err = OS::get_singleton()->execute("/usr/bin/xcrun", args, &r_pipe, &exit_code, true);
		if (!r_pipe.is_empty()) {
			print_line(r_pipe);
		}
		if (err != OK) {
			ERR_PRINT(vformat("Metal compiler returned error code: %d", err));
		}

		if (exit_code != 0) {
			ERR_PRINT(vformat("Metal compiler exited with error code: %d", exit_code));
		}
		int len = result_file->get_length();
		ERR_FAIL_COND_V_MSG(len == 0, ERR_CANT_CREATE, "Metal compiler created empty library");
	}
```

2. ``-mios-version-min=X``

```objc
	// Build the metallib binary.
	{
		List<String> args{ "-sdk", sdk, "metal", "-O3" };

		// Compile metal shaders for the minimum supported target instead of the host machine
		if (export_mode && !min_os_version.is_empty()) {
			switch (device_profile->platform) {
				case MetalDeviceProfile::Platform::macOS: {
					args.push_back("-mmacosx-version-min=" + min_os_version);
					break;
				}
				case MetalDeviceProfile::Platform::iOS: {
					args.push_back("-mios-version-min=" + min_os_version);
					break;
				}
			}
		}

		if (p_stage_data.is_position_invariant) {
			args.push_back("-fpreserve-invariance");
		}
		args.push_back("-fmetal-math-mode=fast");
		args.push_back(source_file->get_path_absolute());
		args.push_back("-o");
		args.push_back(result_file->get_path_absolute());
		String r_pipe;
		int exit_code;
		Error err = OS::get_singleton()->execute("/usr/bin/xcrun", args, &r_pipe, &exit_code, true);
		if (!r_pipe.is_empty()) {
			print_line(r_pipe);
		}
		if (err != OK) {
			ERR_PRINT(vformat("Metal compiler returned error code: %d", err));
		}

		if (exit_code != 0) {
			ERR_PRINT(vformat("Metal compiler exited with error code: %d", exit_code));
		}
		int len = result_file->get_length();
		ERR_FAIL_COND_V_MSG(len == 0, ERR_CANT_CREATE, "Metal compiler created empty library");
	}
```

3. ``-mios-version-min=X`` + ``-std=metalX.Y``

```objc
	// Build the metallib binary.
	{
		List<String> args{ "-sdk", sdk, "metal", "-O3" };

		// Compile metal shaders for the minimum supported target instead of the host machine
		if (export_mode && !min_os_version.is_empty()) {
			switch (device_profile->platform) {
				case MetalDeviceProfile::Platform::macOS: {
					args.push_back("-mmacosx-version-min=" + min_os_version);
					break;
				}
				case MetalDeviceProfile::Platform::iOS: {
					args.push_back("-mios-version-min=" + min_os_version);
					break;
				}
			}
		}

		args.push_back("-std=metal3.0");

		if (p_stage_data.is_position_invariant) {
			args.push_back("-fpreserve-invariance");
		}
		args.push_back("-fmetal-math-mode=fast");
		args.push_back(source_file->get_path_absolute());
		args.push_back("-o");
		args.push_back(result_file->get_path_absolute());
		String r_pipe;
		int exit_code;
		Error err = OS::get_singleton()->execute("/usr/bin/xcrun", args, &r_pipe, &exit_code, true);
		if (!r_pipe.is_empty()) {
			print_line(r_pipe);
		}
		if (err != OK) {
			ERR_PRINT(vformat("Metal compiler returned error code: %d", err));
		}

		if (exit_code != 0) {
			ERR_PRINT(vformat("Metal compiler exited with error code: %d", exit_code));
		}
		int len = result_file->get_length();
		ERR_FAIL_COND_V_MSG(len == 0, ERR_CANT_CREATE, "Metal compiler created empty library");
	}
```

### xcrun results

| Command | Result |
| - | - |
| ``-std=metal2.6`` | ✅ |
| ``-std=metal2.7`` | ✅ |
| ``-std=metal2.8`` | ❔ |
| ``-std=metal2.9`` | ❔ |
| ``-std=metal3.0`` | ❔ |
| ``-std=metal3.1`` | ❌ |
| ``-std=metal3.2`` | ❌ |
| ``-mios-version-min=14`` | ✅ |
| ``-mios-version-min=17`` | ✅ |
| ``-mios-version-min=18`` | ✅ (wtf 👀) |
| ``-mios-version-min=14`` + ``-std=metal3.1`` | ✅ |

I plan on testing out those other ``-std=x.x`` commands when I get access to the device again tomorrow.  Because I don't have access to apple's source code, I can only theorize as to what is happening based on the outputs.  I'm feeling quite confident on some things.  I believe the ``-mios-version-min`` arg targets a metal version even lower than the actual min.  I suspect the ``-mios-version-min=18`` arg is compiling ``metallib`` against a ``metal2.x`` version.  This would explain why targetting this min os still runs on 17.6.1.  And when combining both ``-mios-version-min=`` and ``-std=`` args, I suspect ``-std=`` is completely ignored.

Now why doesn't ``-std=metal3.1`` work on ``iOS 17.6.1 A15 GPU (Apple8)``?  I have no idea.  This doesn't make sense to me, but I've tested this with multiple variations.  I even changed the target OS to ``17.6.1`` in the xcodeproj AND in the ``export_preset``.  Both still don't work for ``metal3.1``.

Apple's official docs say it should, but these docs are not holding true on this device.

So what should we do as the official solution here in Godot?  I'm leaning towards a couple possibilities.  And we could even let the user decide based on export preset settings.

1. Target a metal 2.x version for all iOS versions less than iOS 18.  If iOS is greater than 18, we can target metal 3.x
2. Pass ``-mios-version-min=`` to ``xcrun`` and let ``xcrun`` make that decision for us.  Unfortanetely, the tests indicate to me that ``xcrun`` is extremely conservative in targeting a low metal version when this is passed.  This may frustrate users who specifically want a metal3 or metal4 version for certain features.
3. Expose a ``metal_version`` config in ``export_preset``.  If this is set, pass ``std=`` to xcrun.  Otherwise, do either:
    - ``-mios-version-min=``
    - refer to option 1 above

clayjohn commented:
Thanks for the great work!

Closing as superseded by https://github.com/godotengine/godot/pull/110264
